Lib/test/test_httpservers.py

```diff
@@ -315,6 +360,70 @@ def test_head_via_send_error(self):
             self.assertEqual(b'', data)
 
 
+def certdata_file(*path):
+    return os.path.join(os.path.dirname(__file__), "certdata", *path)
+
+
+@unittest.skipIf(ssl is None, "requires ssl")
+class BaseHTTPSServerTestCase(BaseTestCase):
+    CERTFILE = certdata_file("keycert.pem")
+    ONLYCERT = certdata_file("ssl_cert.pem")
+    ONLYKEY = certdata_file("ssl_key.pem")
+    CERTFILE_PROTECTED = certdata_file("keycert.passwd.pem")
+    ONLYKEY_PROTECTED = certdata_file("ssl_key.passwd.pem")
+    EMPTYCERT = certdata_file("nullcert.pem")
+    BADCERT = certdata_file("badcert.pem")
+    KEY_PASSWORD = "somepass"
+    BADPASSWORD = "badpass"
+
+    tls = (ONLYCERT, ONLYKEY, None)  # values by default
+
+    request_handler = DummyRequestHandler
+
+    def test_get(self):
+        response = self.request('/')
+        self.assertEqual(response.status, HTTPStatus.OK)
+
+    def request(self, uri, method='GET', body=None, headers={}):
+        context = ssl._create_unverified_context()
+        self.connection = http.client.HTTPSConnection(
+            self.HOST, self.PORT, context=context
+        )
+        self.connection.request(method, uri, body, headers)
+        return self.connection.getresponse()
+
+    def test_valid_certdata(self):
+        valid_certdata= [
+            (self.CERTFILE, None, None),
+            (self.CERTFILE, self.CERTFILE, None),
+            (self.CERTFILE_PROTECTED, None, self.KEY_PASSWORD),
+            (self.ONLYCERT, self.ONLYKEY_PROTECTED, self.KEY_PASSWORD),
+        ]
+        for certfile, keyfile, password in valid_certdata:
+            with self.subTest(
+                certfile=certfile, keyfile=keyfile, password=password
+            ):
+                server = create_https_server(certfile, keyfile, password)
+                self.assertIsInstance(server, HTTPSServer)
+                server.server_close()
+
+    def test_invalid_certdata(self):
+        invalid_certdata = [
+            (self.BADCERT, None, None),
+            (self.EMPTYCERT, None, None),
+            (self.ONLYCERT, None, None),
+            (self.ONLYKEY, None, None),
+            (self.ONLYKEY, self.ONLYCERT, None),
```

Comment:
There is one test we need to add when we specify a cert and a password-protected file, but no password. By using `load_cert_chain`, the docs say:

> If the password argument is not specified and a password is required, OpenSSL’s built-in password prompting mechanism will be used to interactively prompt the user for a password.

So there would be an *interactive* prompt somewhere. I'm not sure we can disable this though. Can you check that we don't hang?

The test case would be `(self.CERTFILE_PROTECTED, None, None)`.

Reply:
thank you. really good case but I don't know how to check this case, especially since the hang will affect the execution time of the tests
I also checked `test_ssl` which checks for something like this and they don't have a test for this either

Reply:
This may be an issue where we should add a new parameter that disables password prompt for instance. However, this requires to modify the SSL module (in a separate PR).

@gpshead How do we deal with this situation?

Reply:
IMO, PR looks ready if we want to add this test case we can discuss it separately or ping someone else

Reply:
Gregory is however the one who we should ping though. Now, since `test_ssl` doesn't seem to handle that case, we may discuss it separately. Can you open an issue for that? (and maybe provide an example where we would have the password prompt) Then add some `TODO` here to indicate the possible issue we have so that we don't forget.

Reply:
interactive coming out of a third party library?  we don't need to test their functionality.  if it doesn't work and someone wants it to work i'm sure we'll get a bugreport.

Reply:
>  we don't need to test their functionality. if it doesn't work and someone wants it to work i'm sure we'll get a bugreport

I was more worried about how we actually use it, so we created https://github.com/python/cpython/issues/132102. But if you think it's fine not to cover this, I'm also fine (we test everything but the interactive prompt; but I guess that the interactive prompt is actually populating the context so we don't need to do anything on our side).